### ISSUE TITLE
System.IndexOutOfRangeException from App After you upgraded SC

### ISSUE THREAD

**Issue by andwah** [Starcounter/Starcounter#2487]
I upgraded Starcounter (Polyjuice Edition) from a version from mid December last yet to latest (today).

When i start the App "Launcher" (but i guess any app will do). i get the following Exception 

```
[14:05:44, Starcounter (calendar)]
System.IndexOutOfRangeException: Index was outside the bounds of the array.
   at Starcounter.Binding.TypeDef.PopulatePropertyDef(TypeDef[] typeDefs) in c:\TeamCity\TeamCity8\buildAgent\work\sc-2.0.2991.3\Level1\src\Starcounter\Binding\TypeDef.cs:line 170
   at Starcounter.Hosting.Package.UpdateDatabaseSchemaAndRegisterTypes(TypeDef[] unregisteredTypeDefs) in c:\TeamCity\TeamCity8\buildAgent\work\sc-2.0.2991.3\Level1\src\Hosting\Starcounter.Hosting\Package.cs:line 263
   at Starcounter.Hosting.Package.ProcessWithinCurrentApplication(Application application) in c:\TeamCity\TeamCity8\buildAgent\work\sc-2.0.2991.3\Level1\src\Hosting\Starcounter.Hosting\Package.cs:line 193
   at Starcounter.Hosting.Package.Process() in c:\TeamCity\TeamCity8\buildAgent\work\sc-2.0.2991.3\Level1\src\Hosting\Starcounter.Hosting\Package.cs:line 125
   at Starcounter.Hosting.Package.Process(IntPtr hPackage) in c:\TeamCity\TeamCity8\buildAgent\work\sc-2.0.2991.3\Level1\src\Hosting\Starcounter.Hosting\Package.cs:line 56
   at StarcounterInternal.Hosting.Processor.RunMessageLoop(Void* hsched) in c:\TeamCity\TeamCity8\buildAgent\work\sc-2.0.2991.3\Level1\src\Hosting\Starcounter.Hosting\Processor.cs:line 66
HResult=-2146233080
```

The used database and starcounter binaries is in your dropbox.

**Comment by k-rus** [comment]
According to @andwah database schema, which is SO in this case, was updated at the same time.

I don't know if it can be related to private or code properties. I will investigate.

**Comment by malx122** [comment]
Yepp, this has to do with SO. We have the same problem trying to go from our SO to the polyjuice edition.

**Comment by per-samuelsson** [comment]
Seems that the problem here and in #2485 might be related in some way. They both occur when binding the new SO-libraries (in the Polyjuice edition). My guess it's something in there we can't handle properly, and that it might be related to private members.

**Comment by k-rus** [comment]
> Seems that the problem here and in #2485 might be related in some way. They both occur when binding the new SO-libraries (in the Polyjuice edition). My guess it's something in there we can't handle properly, and that it might be related to private members.

I found reported in #2485 that it is related to a protected property, which is not handled correctly during binding, and I agree this issue should be related. I assign this issue to @per-samuelsson too.

**Comment by k-rus** [comment]
I repeat what I wrote in #2485:

>We discussed the issue. The work around is to fix SO that it does not define any private/protected database properties or fields, since it is not handle in the current version of Starcounter. This has to be changed back when new Starcounter version, currently *next*, is in place. I guess @andwah will do work around and then the issue will be placed in backlog to test with *next* version in future.

**Comment by andwah** [comment]
I have now pushed a fix into Society-Objects repo.
I changed all non public properties to public properties.
(it will be in the next build of the Polyjuice Edition)
see #2485

**Comment by malx122** [comment]
This does not solv #2487

**Comment by malx122** [comment]
I still get the IndexOutOfRangeException. 
The problem is probably that our "old" databases includes a SO with no property "Type" on ring1.Something. Instead we had different properties "Type" on ring8 classes. When changing to polyjuice SO we had to change name in our ring8 to avoid conflicts. This might be the problem when trying to load old databases into Starcounter polyjuice edition

You have no need for our application, nor our code to test this. All that is required is a copy of one of the failing databases, I can provide that.

**Comment by k-rus** [comment]
> I still get the IndexOutOfRangeException.

Is it the same stack trace?



> The problem is probably that our "old" databases includes a SO with no property "Type" on ring1.Something. Instead we had different properties "Type" on ring8 classes. When changing to polyjuice SO we had to change name in our ring8 to avoid conflicts. This might be the problem when trying to load old databases into Starcounter polyjuice edition

Thank you for the explanation! This is a complex case, which, I guess, current version of Starcounter cannot handle. We will investigate what the problem is exactly.



> You have no need for our application, nor our code to test this. All that is required is a copy of one of the failing databases, I can provide that.

Providing the database will be useful. Iesn't it required `Ring8` to reproduce the issue? Can you describe how to reproduce the issue?

As work around have you tried to rename `Type` in `Ring8` using old version of SO and then upgrade to the newest version of SO?
Can you try the work around and report? We will discuss this issue and will report here how we are going to proceed.

**Comment by malx122** [comment]
@k-rus , please contact me on hangout so I can help you reproduce the issue!

**Comment by k-rus** [comment]
The version for the application is from December 22.

**Comment by k-rus** [comment]
The suggested approach for work around is:

1. Get the application with the database running on Starcounter (non-Polyjuice version) from the 16th of January or earlier.
2. Remove property Type and test that the application is working.
3. Unload the database
4. Install latest Starcounter with Polyjuice edition
5. Build and start the application without data
6. Reload the data

**Comment by k-rus** [comment]
Immediate tasks in this issue:

1. Reproduce issue on a small sample.
2. Fix error message to give more meaningful information.

Long term solutions are either auto-fixing meta-data, issue #2460, or unload/reload request during installation, issue #2496.

**Comment by parherman** [comment]
We have reproduced the issue in a VM. Steps:

1. Build Level1 Debug with VS2012
2. Add the database from dropbox
3. Start the database in the admin interface

It will fail with the exception in this issue.

**Comment by k-rus** [comment]
Thanks for reproducing the issue!

This error is due to changes in metadata schema. The work around is to do unload from the old database and reload into empty database with application schema.

This is similar to #2495 and solution is described in [this comment](https://github.com/Starcounter/Starcounter/issues/2495#issuecomment-70645979):

Immediate tasks in this issue:

- Reproduce issue on a small sample.
- Fix error message to give more meaningful information.

Long term solutions are either auto-fixing meta-data, issue #2460, or unload/reload request during installation, issue #2496.

**Comment by k-rus** [comment]
> Reproduce issue on a small sample.

The reproduction requires to use right old and then new version. So it might be not so easy to reproduce.



> Fix error message to give more meaningful information.

This will be done in this issue after #2550.



> Long term solutions are either auto-fixing meta-data

This will not help in this case, since it is a lot of work to maintain auto-fixing between all versions of meta-data.


> unload/reload request during installation, issue #2496.

This is valid and will be done in its own issue after #2550.

**Comment by k-rus** [comment]
The code with error is being refactorred in pnext and it seems that this problem was not seeing any more. So I close issue as obsolete and wouldn't fix.
